Mathlib/NumberTheory/IntervalSums.lean

```diff
@@ -0,0 +1,182 @@
+Authors: Chris Birkbeck
+-/
+
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Algebra.Order.Archimedean.Basic
+import Mathlib.Algebra.CharP.Defs
+import Mathlib.Topology.Algebra.Monoid.Defs
+import Mathlib.Topology.Algebra.InfiniteSum.Defs
+import Mathlib.Data.Int.Interval
+import Mathlib.Order.Filter.AtTopBot.Finset
+import Mathlib.Order.Filter.AtTopBot.Group
+
+/-!
+# Sums over symmetric integer intervals
+
+This file contains some lemmas about sums over symmetric integer intervals `Ixx -N N` used in the
+definition of the Eisenstein series `E2`. In particular we define `SymmetricConditional`,
+`IcoFilter`, `IocFilter` as `SummationFilter`s  corresponding to the intervals
+`Icc -N N`, `Ico -N N`, `Ioc -N N` respectively.
+We also prove that these filters are all `NeBot` and `LeAtTop`.
+
+-/
+
+open Finset Topology
+
+--Should this go elsewhere?
+lemma Finset.Icc_succ_succ (n : ℕ) : Icc (-(n + 1) : ℤ) (n + 1) = Icc (-n : ℤ) n ∪
+  {(-(n + 1) : ℤ), (n + 1 : ℤ)} := by
+  refine Finset.ext_iff.mpr ?_
+  intro a
+  simp only [neg_add_rev, Int.reduceNeg, mem_Icc, add_neg_le_iff_le_add, union_insert,
+    union_singleton, mem_insert]
+  omega
+
+lemma Finset.sum_Icc_of_even_eq_range {α : Type*} [CommRing α] {f : ℤ → α} (hf : ∀ n, f n = f (-n))
+    (N : ℕ) : ∑ m ∈ Icc (-N : ℤ) N, f m = 2 * ∑ m ∈ range (N + 1), f m - f 0 := by
+  induction N with
+  | zero => simp [two_mul]
+  | succ N ih =>
+    have := Icc_succ_succ N
+    simp only [neg_add_rev, Int.reduceNeg, Nat.cast_add, Nat.cast_one] at *
+    rw [this, sum_union (by simp), sum_pair (by omega), ih]
+    nth_rw 2 [sum_range_succ]
+    grind
+
+@[to_additive]
+lemma Finset.prod_Icc_eq_prod_Ico_succ {α : Type*} [CommMonoid α] (f : ℤ → α) {l u : ℤ}
+    (h : l ≤ u) : ∏ m ∈ Icc l u, f m = (∏ m ∈ Ico l u, f m) * f u := by
+  simp only [Icc_eq_cons_Ico h, cons_eq_insert, mem_Ico, lt_self_iff_false, and_false,
+    not_false_eq_true, prod_insert, mul_comm]
+
+lemma Finset.sum_Icc_add_endpoints {R : Type*} [AddCommGroup R] (f : ℤ → R) {N : ℕ} (hn : 1 ≤ N) :
+    ∑ m ∈ Icc (-N : ℤ) N, f m = f N + f (-N : ℤ) + ∑ m ∈ Icc (-(N - 1) : ℤ) (N - 1), f m := by
+  induction N
+  · grind
+  · zify
+    rw [Icc_succ_succ, sum_union (by simp)]
+    grind
+
+section IntervalFilters
+
+open Filter Function Finset SummationFilter
+
+variable (G : Type*) [Neg G] [Preorder G] [LocallyFiniteOrder G]
+
+/-- The SummationFilter on Locally finite order `G` corresponding to the symmetric
+intervals `Icc (-N) N`· -/
+def SummationFilter.symmetricConditional :
+    SummationFilter G where
+  filter := atTop.map (fun g ↦ Icc (-g) g)
+
+/-- The SummationFilter on `ℤ` corresponding to the intervals `Icc -N N`. Note that this is
+the same as the limit over open intervals `Ioo -N N` (see `SymmetricConditional_eq_map_Ioo`). -/
+abbrev SummationFilter.symCondInt : SummationFilter ℤ :=
+  SummationFilter.symmetricConditional ℤ
+
+lemma SymmetricConditional_eq_map_Icc :
```

Comment:
Can you do this in the full generality that `SummationFilter.symmetricConditional` is defined? Its name should be lowerCamelCased (it's a lemma), and it should probably be in the `SummationFilter` namespace.